### ISSUE TITLE
git handling skips submodules

### ISSUE THREAD

**Issue by sasq64** [srstevenson/vim-picker#55]
Using `git ls-files --others` is not ideal since it does not include git submodules.

There should probably be an option to ignore the git check and just use `fd` ...

**Comment by sasq64** [comment]
Also, to make matters worse, `git ls-files` shows multiple copies of files during merge operations with conflicts.

**Comment by boxofrox** [comment]
Out of curiosity, does `git ls-files --cached --exclude-standard --others --recurse-submodules` include git submodules?

Perhaps a variable, `g:picker_git_list_flags`, could be used to override the default `ls-files --cached --exclude-standard --others`, and if the variable is empty, just fallback to `g:picker_find_executable` or `find`?

**Comment by sasq64** [comment]
`--recurse-submodules` can't be used with other flags except for `--cached`.

Yes, a variable like that would do the trick.

Also, I'm wondering what the main idea behind this features is? Is it just an optimization, because `fd` can be slow for very large directories?
Because with the `--others` argument you basically get the same result as with `fd`; all files except ignored ones according to _.gitignore_

**Comment by srstevenson** [comment]
Thanks for reporting these issues @sasq64, this helps us make vim-picker better for everyone!

> Using git ls-files --others is not ideal since it does not include git submodules.

This sounds like something that could be reported upstream to the Git maintainers so the commands @boxofrox asked about would work.

> Also, to make matters worse, git ls-files shows multiple copies of files during merge operations with conflicts.

We could pipe the `git ls-files` output to `sort -u` to mitigate this: we already do this with the output from `ctags`.

> Also, I'm wondering what the main idea behind this features is? Is it just an optimization, because fd can be slow for very large directories?

vim-picker predates `fd`, and so didn't originally use it. The main reason for `fd` support is for use outside Git repositories, where it gives more far more useful output than `find`. I didn't drop support for `git ls-files` when I added support for `fd` because I expect fewer people have `fd` installed than Git.

Thinking about how to resolve this, I remembered some thoughts I'd previously had about supporting other version control systems (most notably Mercurial). I've opened #56 with a proposed solution, which allows the user to specify a custom file listing tool which will be used in place of Git, `fd`, and `find`. In the general case this can be a script containing arbitrary logic, such as choosing a different file listing command on network file systems and local disks, or supporting company-internal version control systems. This is both more flexible and less of a maintenance burden than implementing this kind of logic in vim-picker itself. What do people think about #56?
